src/core/remio.c

```diff
@@ -0,0 +1,709 @@
+ */
+static bool remio_handle_rw(unsigned long value, unsigned long cpu_id, unsigned long vcpu_id)
+{
+    if (remio_get_request_state(cpu_id, vcpu_id) != REMIO_STATE_PROCESSING) {
+        return false;
+    }
+    remio_set_request_value(cpu_id, vcpu_id, value);
+    remio_set_request_state(cpu_id, vcpu_id, REMIO_STATE_COMPLETE);
+    return true;
+}
+
+/**
+ * @brief Performs the post work after the completion of the I/O request
+ * @note This function is executed by the frontend VM and is responsible for updating the
+ *       vCPU register in case of a read operation and activating the frontend vCPU
+ * @param event Message event (REMIO_CPU_MSG_*)
+ * @param cpu_id CPU ID of the frontend VM that issued the I/O request
+ * @param vcpu_id vCPU ID of the frontend VM that issued the I/O request
+ * @return Returns true if the operation was successful, false otherwise
+ */
+static bool remio_cpu_post_work(uint32_t event, uint8_t cpu_id, uint8_t vcpu_id)
+{
+    if (remio_get_request_state(cpu_id, vcpu_id) != REMIO_STATE_COMPLETE) {
+        return false;
+    }
+    struct remio_request* request = remio_get_request(cpu_id, vcpu_id);
+
+    switch (event) {
+        case REMIO_CPU_MSG_READ:
+            vcpu_writereg(cpu()->vcpu, request->reg, request->value);
+            break;
+        default:
+            break;
+    }
+
+    remio_set_request_state(cpu_id, vcpu_id, REMIO_STATE_FREE);
+
+    cpu()->vcpu->active = true;
+
+    return true;
+}
+
+long int remio_hypercall(unsigned long arg0, unsigned long arg1, unsigned long arg2)
+{
+    long int ret = -HC_E_SUCCESS;
+    unsigned long virt_remio_dev_id = arg0;
+    unsigned long addr = arg1;
+    unsigned long op = arg2;
+    unsigned long value = vcpu_readreg(cpu()->vcpu, HYPCALL_IN_ARG_REG(3));
+    unsigned long cpu_id = vcpu_readreg(cpu()->vcpu, HYPCALL_IN_ARG_REG(4));
+    unsigned long vcpu_id = vcpu_readreg(cpu()->vcpu, HYPCALL_IN_ARG_REG(5));
+    struct remio_device* device = NULL;
+    struct vm* vm = cpu()->vcpu->vm;
+
+    /** Check if the virtual Remote I/O device ID is within the valid range */
+    if (virt_remio_dev_id >= vm->remio_dev_num) {
+        ERROR("Remote I/O ID (%d) exceeds the valid range for backend VM (%d)", virt_remio_dev_id,
+            vm->id);
+        return -HC_E_FAILURE;
+    }
+
+    /** Get the Remote I/O device ID based on the virtual Remote I/O device ID */
+    unsigned long remio_dev_id = vm->remio_devs[virt_remio_dev_id].id;
+
+    /** Find the Remote I/O device associated with the current backend VM */
+    device = remio_find_vm_dev_by_id(vm, remio_dev_id);
+    if (device == NULL) {
+        ERROR("The Remote I/O backend device (%d) is not associated with the current backend VM "
+              "(%d)",
+            virt_remio_dev_id, vm->id);
+        return -HC_E_FAILURE;
+    }
+
+    switch (op) {
+        case REMIO_HYP_WRITE:
+        case REMIO_HYP_READ:
+            if (!remio_handle_rw(value, cpu_id, vcpu_id)) {
```

Comment:
But this is not checked on the `remio_handle_rw` call. The `cpu_id`/`vcpu_id` are used to access the request without any check.

Reply:
Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/3e9940ee1b0566c3ad0b9049454f3e61f7adbc1b
After validating and extracting the associated Remote I/O device from the backend VM, we now pass the device pointer to the `remio_handle_rw`. Since the requests are tied to the specific Remote I/O device, this ensures that only the correct device on the correct backend VM can modify its state.

Reply:
There is still an issue here. Calling `remio_handle_rw` will end up in calling `remio_get_request` that performs the access to the request array without checking if the id is valid!

Reply:
Done [here](https://github.com/bao-project/bao-hypervisor/pull/176/commits/77ff8e893f35c268097812e29bcc0463c8796d43)